src/sinks/file/mod.rs

```diff
@@ -88,98 +98,125 @@ impl PartitionedFileSink {
 
         Some(bytes)
     }
-}
-
-impl Sink for PartitionedFileSink {
-    type SinkItem = Event;
-    type SinkError = ();
 
-    fn start_send(&mut self, event: Self::SinkItem) -> StartSend<Self::SinkItem, Self::SinkError> {
-        if let Some(key) = self.partition_event(&event) {
-            let encoding = self.encoding.clone();
-
-            let partition = self
-                .partitions
-                .entry(key.clone())
-                .or_insert_with(|| File::new(key.clone(), encoding.clone()));
-
-            self.last_accessed.insert(key.clone(), Instant::now());
-
-            partition
-                .start_send(event)
-                .map_err(|error| error!(message = "Error writing to partition.", %error))
-        } else {
-            Ok(AsyncSink::Ready)
-        }
+    fn deadline_at(&self) -> Instant {
+        Instant::now()
+            .checked_add(self.idle_timeout)
+            .expect("unable to compute next deadline")
     }
 
-    fn poll_complete(&mut self) -> Poll<(), Self::SinkError> {
-        let mut all_files_ready = false;
-
-        for (file, partition) in &mut self.partitions {
-            let ready = match partition.poll_complete() {
-                Ok(Async::Ready(())) => true,
-                Ok(Async::NotReady) => false,
-                Err(error) => {
-                    let file = String::from_utf8_lossy(&file[..]);
-                    error!(message = "Unable to flush file.", %file, %error);
-                    true
+    async fn run(&mut self, mut input: Receiver<Event>) -> crate::Result<()> {
+        loop {
+            let what = select(input.next(), self.files.next()).await;
```

Comment:
Can we use the new tokio select here?

Reply:
I haven't used `tokio::select`, but I've used `futures::select` before, and I try to keep away from them until they're absolutely necessary. The reason for that is they're huge macros, and significantly complicate the debugging process and other kinds of troubleshooting. Using `select` fn, on the other hand, is simple, and is only a minor syntax annoyance compared to a headache of dealing with complicated macros.
I'd prefer if we leave it as-is :smile:

Reply:
Futures select is hacky while tokio's is much better and isn't 100% a proc macro, I'd suggest trying it out because the Either's everywhere is super hard to read :)

Reply:
I gave it a try, hopefully it'll be better.

As a side effect, I had to switch to nested matches, and now the nesting is real deep.
Before: `loop` > `match` (two levels)
After: `loop` > `tokio::select!`(two levels) -> `match` (two levels)

It's not critical, but is there a way to work around this?

Reply:
It works, so I'll leave it for now.